db/similarity.py

```diff
@@ -0,0 +1,144 @@
+from __future__ import absolute_import
+
+import db
+from db.data import count_all_lowlevel
+from db.exceptions import NoDataFoundException, BadDataException
+import similarity.metrics
+import similarity.utils
+
+from sqlalchemy import text
+
+
+def add_metrics(batch_size):
+    """Computes each metric from the similarity_metrics table for
+    each recording in the lowlevel table, in batches.
+    """
+    lowlevel_count = count_all_lowlevel()
+
+    with db.engine.connect() as connection:
+        metrics = similarity.utils.init_metrics()
+        sim_count = count_similarity()
+        print("Processed {} / {} ({:.3f}%)".format(sim_count,
+                                                   lowlevel_count,
+                                                   float(sim_count) / lowlevel_count * 100))
+        
+        batch_query = text("""
+            SELECT ll.id
+                 , llj.data AS ll_data
+                 , jsonb_object_agg(hlm.model, hlm.data) as hl_data
+              FROM (
+            SELECT id 
+              FROM lowlevel 
+         LEFT JOIN similarity 
+             USING (id) 
+             WHERE similarity.id is NULL) ll
+              JOIN lowlevel_json AS llj USING (id)
+              JOIN highlevel_model AS hlm USING (id) 
+          GROUP BY (ll.id, llj.data)
+             LIMIT :batch_size
+        """)
+
+        while True:
+            with connection.begin():
+                result = connection.execute(batch_query, {"batch_size": batch_size})
+                if not result.rowcount:
+                    break
+
+                row = result.fetchone()
```

Comment:
I had an issue here that was unexpected, using result.fetchall() gave odd behaviour where it took very long to fetch, then eventually timed out (or something like that), and the only way to escape it seemed to be to close the terminal session and restart docker. I thought this might be because of the size of the rows? Fetching one row at a time works without the issue.